File: tools/agent-app/src/main/cliSession.js
'use strict';

// Native CLI receipts live on the tmux host, so the same path works after an app
// restart and when Electron is on Windows while the CLI runs in WSL.
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { parse } = require('smol-toml');
const host = require('./host');
const supported = (cli) => cli === 'codex' || cli === 'kiro';

function runtimePath({ home, id, cli }) {
  if (!/^[a-zA-Z0-9_-]+$/.test(id) || !/^[a-zA-Z0-9_-]+$/.test(cli)) throw new Error('セッション記録先が不正です');
  return host.joinHost(home, `.local/state/agent-app/cli-sessions/${id}/${cli}`);
}

async function read(options) {
  if (!supported(options.cli)) return '';
  const result = await options.shell.exec(['cat', host.joinHost(runtimePath(options), 'session.json')]);
  if (!result.ok) return '';
  try {
    const value = JSON.parse(result.output).id;
    return typeof value === 'string' && /^[a-zA-Z0-9_.:-]{1,200}$/.test(value) ? value : '';
  } catch { return ''; }
}

async function codexNotify({ shell, home, env = {}, argv }) {
  const root = env.CODEX_HOME || (await shell.run('printf %s "${CODEX_HOME:-}"')).output || host.joinHost(home, '.codex');
  async function config(file) {
    const result = await shell.exec(['cat', file]);
    return result.ok ? parse(result.output) : {};
  }
  const base = await config(host.joinHost(root, 'config.toml'));
  let value = base.notify;
  let profile = base.profile || '';
  for (let i = 0; i < argv.length; i += 1) {
    if (['--profile', '-p'].includes(argv[i])) profile = argv[i + 1] || '';
    if (argv[i].startsWith('--profile=')) profile = argv[i].slice(10);
  }
  if (profile) {
    if (!/^[a-zA-Z0-9_.-]+$/.test(profile)) throw new Error('Codex のプロファイル名が不正です');
    const external = await config(host.joinHost(root, `${profile}.config.toml`));
    value = external.notify ?? base.profiles?.[profile]?.notify ?? value;
  }
  for (let i = 0; i < argv.length; i += 1) {
    const arg = ['--config', '-c'].includes(argv[i]) ? argv[i + 1] : argv[i].replace(/^--config=/, '');
    if (/^notify\s*=/.test(arg || '')) value = parse(arg).notify;
  }
  if (value != null && (!Array.isArray(value) || !value.every((item) => typeof item === 'string'))) {
    throw new Error('Codex の通知設定を読み取れません');
  }
  return value || [];
}

async function prepare(options) {
  const { shell, cli, argv, env = {} } = options;
  if (!supported(cli)) return { argv, env };
  const chained = cli === 'codex' ? await codexNotify(options) : [];
  const runtime = runtimePath(options);
  const script = host.joinHost(runtime, 'capture.py');
  const source = fs.readFileSync(path.join(__dirname, 'cli-session.py'), 'utf8');
  const install = await shell.exec(['python3', '-c',
    'import os,sys,pathlib; p=pathlib.Path(sys.argv[1]); p.parent.mkdir(parents=True,exist_ok=True,mode=0o700); p.write_text(sys.argv[2]); p.chmod(0o600)',
    script, source]);
  if (!install.ok) throw new Error(`セッションIDの記録にはホストの Python 3 が必要です: ${install.error}`);
  const result = await shell.exec(['python3', script, 'prepare', JSON.stringify({
    cli, argv, env, cwd: options.cwd, runtime, chained, token: crypto.randomUUID(),
  })]);
  if (!result.ok) throw new Error(`セッションIDの記録を準備できません: ${result.error}`);
  return JSON.parse(result.output);
}

module.exports = { prepare, read, supported };

File: tools/agent-app/src/main/cli-session.py
"""Host-side session receipt. Runs on the CLI host, including WSL."""
import json
import fcntl
from contextlib import contextmanager
import os
from pathlib import Path
import re
import shlex
import subprocess
import sys
import tempfile


def write_json(file, data):
    file = Path(file)
    file.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    fd, name = tempfile.mkstemp(dir=file.parent)
    try:
        with os.fdopen(fd, "w") as stream:
            json.dump(data, stream)
        os.replace(name, file)
    finally:
        if os.path.exists(name):
            os.unlink(name)


@contextmanager
def locked(file):
    with open(str(file) + ".lock", "a") as stream:
        os.chmod(stream.name, 0o600)
        fcntl.flock(stream, fcntl.LOCK_EX)
        yield


def record(file, token, cli, chained, payload):
    try:
        event = json.loads(payload)
        sid = event.get("thread-id") if cli == "codex" else event.get("session_id")
        if cli == "codex" and event.get("type") != "agent-turn-complete":
            return
        with locked(file):
            saved = json.loads(Path(file).read_text())
            if saved.get("token") == token and isinstance(sid, str) and re.fullmatch(r"[a-zA-Z0-9_.:-]{1,200}", sid):
                write_json(file, {"token": token, "id": sid})
    except (OSError, ValueError, TypeError, AttributeError):
        pass
    # A user's existing notification still receives the original payload.
    if chained:
        try:
            subprocess.run([*chained, payload], stdin=subprocess.DEVNULL,
                           stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=10)
        except (OSError, ValueError, subprocess.TimeoutExpired):
            pass


def prepare(opts):
    cli, argv = opts["cli"], list(opts["argv"])
    env = dict(opts.get("env") or {})
    runtime = Path(opts["runtime"])
    runtime.mkdir(parents=True, exist_ok=True, mode=0o700)
    file = runtime / "session.json"
    token = opts["token"]
    hook = [sys.executable, str(Path(__file__).resolve()), "record", str(file), token, cli,
            json.dumps(opts.get("chained") or [])]
    if cli == "codex":
        argv.extend(["--config", "notify=" + json.dumps(hook)])
    elif cli == "kiro":
        if "--v3" in argv or "--agent-engine=v3" in argv:
            raise ValueError("Kiro v3 のセッション記録には未対応です")
        source = Path(env.get("KIRO_HOME") or os.environ.get("KIRO_HOME") or Path.home() / ".kiro")
        private = runtime / "kiro"
        private.mkdir(exist_ok=True, mode=0o700)
        # Keep settings, credentials, skills and the session store at their original paths.
        # Only the generated agent definition belongs to this app conversation.
        if source.is_dir():
            for item in source.iterdir():
                target = private / item.name
                if item.name != "agents" and not target.exists() and not target.is_symlink():
                    target.symlink_to(item.resolve(), target_is_directory=item.is_dir())
        agents = private / "agents"
        agents.mkdir(exist_ok=True, mode=0o700)
        if (source / "agents").is_dir():
            for item in (source / "agents").iterdir():
                target = agents / item.name
                if item.name != "agent-app.json" and not target.exists() and not target.is_symlink():
                    target.symlink_to(item.resolve())
        selected, position = "default", None
        for settings in (source / "settings/cli.json", Path(opts["cwd"]) / ".kiro/settings/cli.json"):
            if settings.is_file():
                selected = json.loads(settings.read_text()).get("chat.defaultAgent") or selected
        for index, arg in enumerate(argv):
            if arg == "--agent" and index + 1 < len(argv):
                selected, position = argv[index + 1], index + 1
            elif arg.startswith("--agent="):
                selected, position = arg.split("=", 1)[1], index
        if not re.fullmatch(r"[a-zA-Z0-9_.-]+", selected):
            raise ValueError("Kiro のエージェント名を読み取れません")
        cwd = Path(opts["cwd"])
        candidates = [cwd / ".kiro" / "agents" / (selected + ".json"), source / "agents" / (selected + ".json")]
        agent_file = next((item for item in candidates if item.is_file()), None)
        if (position is not None or selected != "default") and agent_file is None:
            raise ValueError("Kiro のカスタムエージェントが見つかりません: " + selected)
        agent = json.loads(agent_file.read_text()) if agent_file else {"includeMcpJson": True}
        hooks = agent.setdefault("hooks", {})
        for event in ("agentSpawn", "stop"):
            hooks.setdefault(event, []).append({"command": shlex.join(hook)})
        resources = agent.setdefault("resources", [])
        for uri in ("file://" + str(source / "steering/**/*.md"),
                    "skill://" + str(source / "skills/**/SKILL.md"),
                    "file://" + str(cwd / ".kiro/steering/**/*.md"),
                    "skill://" + str(cwd / ".kiro/skills/**/SKILL.md"),
                    "file://" + str(cwd / "AGENTS.md")):
            if uri not in resources:
                resources.append(uri)
        agent["name"] = "agent-app"
        write_json(agents / "agent-app.json", agent)
        if position is None:
            argv.extend(["--agent", "agent-app"])
        elif argv[position].startswith("--agent="):
            argv[position] = "--agent=agent-app"
        else:
            argv[position] = "agent-app"
        env["KIRO_HOME"] = str(private)
    with locked(file):
        write_json(file, {"token": token, "id": ""})
    return {"argv": argv, "env": env, "file": str(file)}


if __name__ == "__main__":
    if sys.argv[1] == "prepare":
        print(json.dumps(prepare(json.loads(sys.argv[2]))))
    elif sys.argv[1] == "record":
        record(*sys.argv[2:5], json.loads(sys.argv[5]), sys.argv[6] if len(sys.argv) > 6 else sys.stdin.read())
